packages/js/src/core/fetch-token.ts

```diff
@@ -0,0 +1,75 @@
+import { KeysToCamelCase } from '@silverhand/essentials';
+import { string, number, assert, type, StructError, Infer } from 'superstruct';
+import { Except } from 'type-fest';
+
+import { LogtoError } from '../utils/errors';
+import { createRequester } from '../utils/requester';
+
+export interface FetchTokenByRefreshTokenParameters {
+  clientId: string;
+  tokenEndPoint: string;
+  refreshToken: string;
+  resource?: string;
+  scope?: string[];
+}
+
+const TokenResponseSchema = type({
+  access_token: string(),
+  refresh_token: string(),
+  id_token: string(),
+  scope: string(),
+  expires_in: number(),
+});
+
+export type RefreshTokenTokenResponse = Except<
+  KeysToCamelCase<Infer<typeof TokenResponseSchema>>,
+  'scope'
+> & {
+  scope: string[];
+};
+
+export const fetchTokenByRefreshToken = async ({
+  clientId,
+  tokenEndPoint,
+  refreshToken,
+  resource,
+  scope,
+}: FetchTokenByRefreshTokenParameters): Promise<RefreshTokenTokenResponse> => {
+  const parameters = new URLSearchParams();
+  parameters.append('client_id', clientId);
+  parameters.append('refresh_token', refreshToken);
+  if (resource) {
+    parameters.append('resource', resource);
+  }
+
+  if (scope?.length) {
+    parameters.append('scope', scope.join(' '));
+  }
+
+  const requester = createRequester();
+  const response = await requester(tokenEndPoint, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/x-www-form-urlencoded',
+    },
+    body: parameters,
+  });
+
+  try {
+    assert(response, TokenResponseSchema);
+    const { access_token, refresh_token, id_token, scope, expires_in } = response;
+    return {
+      accessToken: access_token,
+      refreshToken: refresh_token,
+      idToken: id_token,
+      scope: scope.split(' '),
+      expiresIn: expires_in,
+    };
+  } catch (error: unknown) {
+    if (error instanceof StructError) {
+      throw new LogtoError('struct_verification.response_struct_mismatched');
```

Comment:
We can let superstruct throw the error directly instead of wrapping another layer to keep simplicity.

Reply:
OK, fine for me. Will update shortly